feature/1: Initial Implementation (#2)
exasol/java-udf-startup-time-improver: README.md

[![Technical Debt](https://sonarcloud.io/api/project_badges/measure?project=com.exasol%3Ajava-udf-startup-time-improver&metric=sqale_index)](https://sonarcloud.io/dashboard?id=com.exasol%3Ajava-udf-startup-time-improver)

[![Code Smells](https://sonarcloud.io/api/project_badges/measure?project=com.exasol%3Ajava-udf-startup-time-improver&metric=code_smells)](https://sonarcloud.io/dashboard?id=com.exasol%3Ajava-udf-startup-time-improver)
[![Coverage](https://sonarcloud.io/api/project_badges/measure?project=com.exasol%3Ajava-udf-startup-time-improver&metric=coverage)](https://sonarcloud.io/dashboard?id=com.exasol%3Ajava-udf-startup-time-improver)
[![Duplicated Lines (%)](https://sonarcloud.io/api/project_badges/measure?project=com.exasol%3Ajava-udf-startup-time-improver&metric=duplicated_lines_density)](https://sonarcloud.io/dashboard?id=com.exasol%3Ajava-udf-startup-time-improver)
[![Lines of Code](https://sonarcloud.io/api/project_badges/measure?project=com.exasol%3Ajava-udf-startup-time-improver&metric=ncloc)](https://sonarcloud.io/dashboard?id=com.exasol%3Ajava-udf-startup-time-improver)

> The Java UDF startup time improver is a tool that optimizes Java UDF. It does so by preloading the Java classes and writing them into a class dump on BucketFS. Then it replaces the script definition with a different one that declares a jvm-option which introduces the JVM to use the dump.

Reviewer: Explain what a UDF is, that it is Exasol specific and add a link to the UDF documentation.
```suggestion
The Java UDF startup time improver is a tool that optimizes Java UDF. It does so by preloading the Java classes and writing them into a class dump on BucketFS. Then it replaces the script definition with a different one that declares a JVM-option which instructs the JVM to use the dump.
```

## Installation

* Download the latest release of this project.
* Upload the jar to BucketFS.

> * Create the following functions:

Reviewer: ```suggestion
1. Download the latest release of this project.
2. Upload the jar to BucketFS.
3. Create the following functions:
```
Add link to documentation how to upload a jar

  ```sql
  CREATE JAVA SCALAR SCRIPT "JAVA_UDF_STARTUP_TIME_IMPROVER" (
    "UDF_DEFINITION" VARCHAR(2000000) UTF8, 
    "CLASSES" VARCHAR(2000) UTF8, 
    "CONNECTION_NAME" VARCHAR(200) UTF8, 
    "BUCKET_FS_PORT" DECIMAL(18,0), 
    "BUCKET_FS_SERVICE" VARCHAR(200) UTF8, 
    "BUCKET_FS_BUCKET" VARCHAR(200) UTF8, 
    "PATH_FOR_DUMP" VARCHAR(200) UTF8)
  RETURNS VARCHAR(2000) UTF8 AS
  %scriptclass com.exasol.udfstartuptimeimprover.UdfStartUpTimeImprover;

>   %jar /buckets/bfsdefault/default/improver.jar;

Reviewer: Maybe rename the jar, e.g. `java-udf-startup-time-improver.jar`. Else users will wonder about this strange jar in the future

  /
  
  CREATE LUA SCRIPT "JAVA_UDF_STARTUP_TIME_IMPROVER" (UDF_SCHEMA,UDF_NAME,CLASSES,CONNECTION_NAME,BUCKET_FS_PORT,BUCKET_FS_SERVICE,BUCKET_FS_BUCKET,PATH_FOR_DUMP) RETURNS ROWCOUNT AS
    local updatedUdfDef = query([[SELECT TEST.JAVA_UDF_STARTUP_TIME_IMPROVER_INT(
      (SELECT SCRIPT_TEXT FROM SYS.EXA_ALL_SCRIPTS WHERE SCRIPT_NAME=:udfName AND SCRIPT_SCHEMA=:udfSchema),:classes,:connection, :bfsPort, :bfsService, :bfsBucket, :pathForDump) AS CMD]], { udfName = UDF_NAME, udfSchema = UDF_SCHEMA, classes = CLASSES, connection = CONNECTION_NAME, bfsPort = BUCKET_FS_PORT, bfsService = BUCKET_FS_SERVICE, bfsBucket = BUCKET_FS_BUCKET, pathForDump = PATH_FOR_DUMP })
    query(updatedUdfDef[1].CMD)
  /
  ```

## Usage

> Crate a connection with the write-password of the BucketFS where the improver can write the class-dump to.

Reviewer: ```suggestion
Create a connection with the write-password of the BucketFS where the improver can write the class-dump to.
```

```sql
CREATE CONNECTION BFS_CONNECTION
TO ''
USER ''
IDENTIFIED BY '<BucketFsWritePassword>';
```

> Download the class list for the UDF / adapter script that you want to optimize. You can find this file attached to the GitHub release. Now upload the file to a bucket in BucketFS.

Reviewer: ```suggestion
Download the class list for the UDF / adapter script that you want to optimize. You can find file `classes.lst` attached to the GitHub release. Now upload the file to a bucket in BucketFS.
```

Would it be possible to include this list in the jar? We already uploaded a jar before.

Author: Nice idea. I'll keep that option in mind and in case it works out, I'll change the guide here.

Run the improver:

```sql
execute script TEST.JAVA_UDF_STARTUP_TIME_IMPROVER(
    '<schema of the script to optimize>', 
    '<name of the script to optimize>', 
    '<path in bucketfs to the listt of classes to preload>', 
    '<name of the connection we just created>', 
    <bucketfs-port>, 
    '<name of the BucketFS service where the improver stors the class-dump>', 
    '<name of the Bucket where the improver stors the class-dump>', 

>     '<path for the class.dump in the bucket>')

Reviewer: ```suggestion
    '<schema of the script to optimize>', 
    '<name of the script to optimize>', 
    '<path in bucketfs to the list of classes to preload>', 
    '<name of the connection we just created>', 
    <bucketfs-port>, 
    '<name of the BucketFS service where the improver stores the class-dump>', 
    '<name of the Bucket where the improver stores the class-dump>', 
    '<path for the class dump file in the bucket>')
```
